# <center>Fake-News-Classifier</center>

![Fake-Fact image](https://theleaflet.in/wp-content/uploads/2020/11/190108-fake-thumb.jpg)

# ♦ Introduction ♦

As of Today, It is very hard to find the difference between the fake and the real news as the social media grows wider.Fake news is false or misleading information presented as news. Fake news often has the aim of damaging the reputation of a person or entity, or making money through advertising revenue.To takle this problem i build a machine learning model which can classify the fake news. It may not be the best model but it is a good model.

#About 
## Dataset:
The following dataset used in this project is the famous dataset is taken from the [kaggle](https://www.kaggle.com/datasets/clmentbisaillon/fake-and-real-news-dataset). The dataset contains two csv files named True.csv and Fake.csv which consists of the real news and fake news respectively. Each csv file contains 4 attributes.

**The Attributes of the Dataset**:
1. Title: 
         Title of the News, containing 17903 unique values
2. Text:
         Text Attribute consists of the news written/published in the articles.
3. Subject:
          The Subject Attribute contains the type of news like Political News, World News or just the regular news.
4. Date:
          The date attribute contains the date that this article was posted at.
        


## Project:
The project **Fake News Classifier** is a part of Big Data  Course Project.This Project is completed under Dr.Kanishka Tyagi.In this Project, I made a model analysis so that anyone can choose the best model according to the accuracy of that model to classify the Fake News.In this Project I used Two Methods in NLP, **Bag of Words** and **TF-IDF**. **Multinomial Naive Bayes** for the Model Training


## The Model is trained on the basis of 'Text of the Article' in the Dataset

lets install all the required libraries to download the dataset

In [56]:
!pip install pandas opendatasets --upgrade --quiet

In [57]:
import pandas as pd
import opendatasets as opd

## Download 

There are several ways to download the dataset and upload it into your jupyter notebook.
Here is one of the method I am using in this project:
1. Install the [opendatasets](https://github.com/JovianML/opendatasets#opendatasets) library into your Notebook.
2. Copy the URL of the dataset and save the URL into a variable.
3. Use the 'download' function to download the csv file of the dataset.
4. Now you can insert the csv file to your jupyter notebook and perform several operations.

Through this process you may asked to enter the username and API key of Kaggle Account, You need to input those values to proceed Downloading.In case it is already Installed, it does not required

Create an API key in Kaggle.

Step 1:  To do this, go to kaggle.com/ and open your user settings page.
![image1](https://i.stack.imgur.com/jxGQv.png)

Step 2:Next, scroll down to the API access section and click generate to download an API key.
![image2](https://i.stack.imgur.com/Hzlhp.png)
This will download a file called kaggle.json to your computer. You'll use this file in Jupyter Notebook/Google Colab to access Kaggle datasets and competitions.

In [58]:
opd.download('https://www.kaggle.com/datasets/clmentbisaillon/fake-and-real-news-dataset')

Skipping, found downloaded files in ".\fake-and-real-news-dataset" (use force=True to force download)


The dataset thas been downloadeed and extracted

# Data Preperation and Cleaning
Lets Load the File 'Fake.csv' into pandas dataframe and perform some cleaning if required.

In [59]:
fake_df = pd.read_csv('./fake-and-real-news-dataset/Fake.csv')

In [60]:
fake_df.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

**.columns()** method give the list of all the attributes presented in the dataset.

In [61]:
fake_df.columns

Index(['title', 'text', 'subject', 'date'], dtype='object')

**.count()** method provides the Total Number of values present in the Dataset

In [62]:
fake_df.count()

title      23481
text       23481
subject    23481
date       23481
dtype: int64

### Checking if there any NULL values present in the Fake Dataframe

**.info()** provides all the information regarding the attributes in the dataset.

In [63]:
fake_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


As there are no NULL values in the datset. let's move on to the next step.

In [64]:
fake_df.memory_usage(deep = True)

Index           128
title       5380518
text       61153828
subject     1507065
date        1646426
dtype: int64

As we are only preparing model using 'title' attribute in the dataset, Here seperating the 'title' column from the dataframe and aperform some operations

In [65]:
df_f = fake_df.loc[ : ,['title']]

In [66]:
df_f.head()

title
0   Donald Trump Sends Out Embarrassing New Year’...
1   Drunk Bragging Trump Staffer Started Russian ...
2   Sheriff David Clarke Becomes An Internet Joke...
3   Trump Is So Obsessed He Even Has Obama’s Name...
4   Pope Francis Just Called Out Donald Trump Dur...

### Checking for the Duplicate Values in the Fake Dataframe

In [67]:
df_f.count()

title    23481
dtype: int64

Here the Total number of values in the dataset is **23481**

Finding out the total number of unique values in the dataset

In [68]:
n = len(pd.unique(df_f['title']))
print(n)

17903


Here the Total Number of Unique Values are **17904** , Remaining **5577** isn't useful for training the model so let's delete them from the dataset

In [69]:
df_f.drop_duplicates(inplace=True)

In [70]:
df_f.count()

title    17903
dtype: int64

Now, The Duplicate values has been removed from the dataset.

In [71]:
df_f.head(5)

title
0   Donald Trump Sends Out Embarrassing New Year’...
1   Drunk Bragging Trump Staffer Started Russian ...
2   Sheriff David Clarke Becomes An Internet Joke...
3   Trump Is So Obsessed He Even Has Obama’s Name...
4   Pope Francis Just Called Out Donald Trump Dur...

Adding a new column named **TorF** to the dataset to identify wheather the news is Fake or True. For this dataset every value is assigned as 0 as every news is Fake. 

In [72]:
df_f['TorF'] = 0

In [73]:
df_f.head(5)

title  TorF
0   Donald Trump Sends Out Embarrassing New Year’...     0
1   Drunk Bragging Trump Staffer Started Russian ...     0
2   Sheriff David Clarke Becomes An Internet Joke...     0
3   Trump Is So Obsessed He Even Has Obama’s Name...     0
4   Pope Francis Just Called Out Donald Trump Dur...     0

Shuffling the Fake Dataframe

In [74]:
df_f = df_f.sample(frac=1).reset_index(drop=True)

In [75]:
df_f.head(10)

title  TorF
0  SAY WHAT?! Muslim Women Can Now Cover Their Fa...     0
1   Republicans Had Total Control Of This State A...     0
2  KARMA: GAY PASTOR SUES WHOLE FOODS For “Anti-G...     0
3  “ENTITLED” DEM REP. SHEILA JACKSON LEE Has Bee...     0
4  WOW! Why George W. Bush Never Left DC Until Af...     0
5   Watergate Reporter Says Trump’s Russia Scanda...     0
6   Chuck Schumer Has A BRILLIANT Response To Tru...     0
7  Breaking: Disrespectful NFL Teams Visiting Lon...     0
8      WIKILEAKS PROVES Hillary Gave Weapons To ISIS     0
9  (VIDEO) REMEMBER THE “OBAMAPHONE”? GET READY T...     0

### Performing the Similar Operations on the True News Dataset

Converting **True.csv** into Pandas Dataframe

In [76]:
True_df = pd.read_csv('./fake-and-real-news-dataset/True.csv')

In [77]:
True_df.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017

In [78]:
True_df.columns

Index(['title', 'text', 'subject', 'date'], dtype='object')

In [79]:
True_df.count()

title      21417
text       21417
subject    21417
date       21417
dtype: int64

In [80]:
True_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB


In [81]:
True_df.describe()

title  \
count                                               21417   
unique                                              20826   
top     Factbox: Trump fills top jobs for his administ...   
freq                                                   14   

                                                     text       subject  \
count                                               21417         21417   
unique                                              21192             2   
top     (Reuters) - Highlights for U.S. President Dona...  politicsNews   
freq                                                    8         11272   

                      date  
count                21417  
unique                 716  
top     December 20, 2017   
freq                   182

In [82]:
True_df.memory_usage(deep = True)

Index           128
title       2609581
text       81025844
subject     1447338
date        1578819
dtype: int64

Extracting the **title** attribute from the dataset and perform data cleaning

In [83]:
df_t = True_df.loc[ : ,['title']]

In [84]:
df_t.head()

title
0  As U.S. budget fight looms, Republicans flip t...
1  U.S. military to accept transgender recruits o...
2  Senior U.S. Republican senator: 'Let Mr. Muell...
3  FBI Russia probe helped by Australian diplomat...
4  Trump wants Postal Service to charge 'much mor...

### Finding out the Total number of duplicate values in the True datsframe

In [85]:
df_t.count()

title    21417
dtype: int64

The Total Number of values in the True Dataframe are **21417**

Finding out the Total Number of Unique Values in the True Dataframe

In [86]:
n = len(pd.unique(df_t['title']))
print(n)

20826


There are **20826** unique values in the True Dataframe. So, there are 591 duplicate values in the dataframe.

Removing the Duplicate Values in the Dataframe

In [87]:
df_t.drop_duplicates(inplace=True)

In [88]:
df_t.count()

title    20826
dtype: int64

In [89]:
df_t.head()

title
0  As U.S. budget fight looms, Republicans flip t...
1  U.S. military to accept transgender recruits o...
2  Senior U.S. Republican senator: 'Let Mr. Muell...
3  FBI Russia probe helped by Australian diplomat...
4  Trump wants Postal Service to charge 'much mor...

Adding a new column **TorF** to differentiate the fake news .

In [90]:
df_t['TorF'] = 1

Here all the values are assigned **1** beacuse the present dataset contains all the true news.

In [91]:
df_t.head()

title  TorF
0  As U.S. budget fight looms, Republicans flip t...     1
1  U.S. military to accept transgender recruits o...     1
2  Senior U.S. Republican senator: 'Let Mr. Muell...     1
3  FBI Russia probe helped by Australian diplomat...     1
4  Trump wants Postal Service to charge 'much mor...     1

Shuffling the Dataframe

In [92]:
df_t = df_t.sample(frac=1).reset_index(drop=True)

In [93]:
df_t.head(10)

title  TorF
0  More than 60 killed in air strike on Syrian ma...     1
1  Brazil's Meirelles jokes about a vice presiden...     1
2  Trump reassures farmers immigration crackdown ...     1
3  At least 29 dead after Islamist attack on Soma...     1
4  Oil mogul Hamm tops Trump list for U.S. energy...     1
5  Australia to move 200 asylum seekers to new PN...     1
6  Lights back on in Venezuela after five-hour bl...     1
7  Clinton's lead over Trump narrows to nine poin...     1
8  Hungary not planning to move Israeli embassy t...     1
9  Egypt attack victims pray, mourn in Sinai mosq...     1

### Combining the two dataframe into one dataframe 

In [94]:
df_combined = pd.concat([df_f,df_t])

In [95]:
df_combined

title  TorF
0      SAY WHAT?! Muslim Women Can Now Cover Their Fa...     0
1       Republicans Had Total Control Of This State A...     0
2      KARMA: GAY PASTOR SUES WHOLE FOODS For “Anti-G...     0
3      “ENTITLED” DEM REP. SHEILA JACKSON LEE Has Bee...     0
4      WOW! Why George W. Bush Never Left DC Until Af...     0
...                                                  ...   ...
20821  Tens of thousands march to defend Hong Kong's ...     1
20822  French conservatives' leader-in-waiting out to...     1
20823  Troops and strays, the only signs of life in r...     1
20824    EU gives formal green light to new Brexit phase     1
20825  Russian and French foreign ministers discussed...     1

[38729 rows x 2 columns]

In [96]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38729 entries, 0 to 20825
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   38729 non-null  object
 1   TorF    38729 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 907.7+ KB


In [97]:
df_combined = df_combined.sample(frac=1).reset_index(drop=True)

In [98]:
df_combined.head(10)

title  TorF
0  Spain's King Felipe says committed to Spanish ...     1
1  Thirty-two years after quake, angry Mexicans s...     1
2  Kremlin says Putin and Trump to discuss North ...     1
3  Trump lawyer: Deutsche Bank has not received s...     1
4  WHY IS LORETTA LYNCH OPENING For One Of The Bi...     0
5   Elijah Cummings Just Asked 5 Questions The Wh...     0
6  Iraq oil ministry warns oil companies against ...     1
7   The Absolutely Cringeworthy Moment Trump Trie...     0
8  No Vacancies: Airbnb shutters Beijing rentals ...     1
9  MILITARY K-9 HANDLER’S Post About Vulgar Encou...     0

# Data pre-processing

The NLP methods using in this Project:
1. Using regular expression, Removing punctuations,special characters and replacing them with spaces abd converting the whole dataframe into lowercase words.
2. Applying Stemming on the whole Dataframe
3. Removing the English Stopwords

The **NLTK** library consists of many useful NLP methods , which we use in this project.

Installing the necessary libraries

In [99]:
import nltk
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

Creating the Object of PorterStemmer() for Stemming.

In [100]:
ps = PorterStemmer()
corpus = []

Iterating each row and applying the regular expressions,removing stopwords and stemming and combining to make a list of sentences.

In [101]:
from  tqdm import tqdm_notebook as tqdm
for i in tqdm(range(0, len(df_combined))):
    review = re.sub('[^a-zA-Z]' , ' ',df_combined['title'][i])
    review = review.lower()
    review = review.split()
    
    review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

C:\Users\rohan\AppData\Local\Temp/ipykernel_14360/1279632879.py:2: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for i in tqdm(range(0, len(df_combined))):


  0%|          | 0/38729 [00:00<?, ?it/s]

Corpus returns a list of all the sentences that are ready to be trained.

In [102]:
corpus

['spain king felip say commit spanish uniti amid catalan crisi',
 'thirti two year quak angri mexican still wait home',
 'kremlin say putin trump discuss north korea vietnam',
 'trump lawyer deutsch bank receiv subpoena trump record',
 'loretta lynch open one biggest black live matter leader endors',
 'elijah cum ask question white hous scare death answer',
 'iraq oil ministri warn oil compani kurdistan contract',
 'absolut cringeworthi moment trump tri flirt irish report video',
 'vacanc airbnb shutter beij rental ahead parti congress',
 'militari k handler post vulgar encount hillari goe viral vote clinton',
 'outspoken lieuten gener name trump top secur advis',
 'trump call mexico pena nieto earthquak wake white hous',
 'unhing dem hijack tucker refus leav show heat argument video',
 'obama put american grave danger new plan close gitmo move prison u',
 'indian court acquit accus g telecom case',
 'caitlyn jenner join celeb mull run u polit offic',
 'easili offend watch american pas

# Bag Of Words

The bag-of-words (BOW) model is a representation that turns arbitrary text into fixed-length vectors by counting how many times each word appears. This process is often referred to as vectorization.

## How is BOW useful?
Despite being a relatively basic model, BOW is often used for Natural Language Processing (NLP) tasks like Text Classification. Its strengths lie in its simplicity: it’s inexpensive to compute, and sometimes simpler is better when positioning or contextual info aren’t relevant.

In [103]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
X = cv.fit_transform(corpus).toarray()

In [104]:
Y = df_combined.iloc[:,1].values

In [105]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y, test_size = 0.20 , random_state = 0)

In [106]:
from sklearn.naive_bayes import MultinomialNB
fake_news_model = MultinomialNB().fit(X_train,Y_train)

In [107]:
Y_pred = fake_news_model.predict(X_test)

In [108]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(Y_test,Y_pred)
matrix

array([[3350,  225],
       [ 310, 3861]], dtype=int64)

In [109]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(Y_test,Y_pred)
accuracy

0.9309320939839918

Here, We got an accuracy of `93`%.This may change because using Random Sampling the highest accuracy i got was `94.4`%, Don't Worry it won't be much Different.

## Saving The Model
Saving Machine Learning Model using joblib.

In [111]:
!pip install joblib --upgrade --quiet
import joblib
filename1 = 'BOW_model.sav'
joblib.dump(fake_news_model, filename1)

['BOW_model.sav']

### To load the data:

loaded_model = joblib.load(filename1)

# Tf-idf

TF-IDF stands for term frequency-inverse document frequency and it is a measure, used in the fields of information retrieval (IR) and machine learning, that can quantify the importance or relevance of string representations (words, phrases, lemmas, etc)  in a document amongst a collection of documents (also known as a corpus).

## Overview of TF-IDF
TF-IDF can be broken down into two parts TF (term frequency) and IDF (inverse document frequency).

### What is TF (term frequency)?
Term frequency works by looking at the frequency of a particular term you are concerned with relative to the document. There are multiple measures, or ways, of defining frequency:

1. Number of times the word appears in a document (raw count).
2. Term frequency adjusted for the length of the document (raw count of occurences divided by number of words in the document).
3. Logarithmically scaled frequency (e.g. log(1 + raw count)).
4. Boolean frequency (e.g. 1 if the term occurs, or 0 if the term does not occur, in the document).

### What is IDF (inverse document frequency)?
Inverse document frequency looks at how common (or uncommon) a word is amongst the corpus. IDF is calculated as follows where t is the term (word) we are looking to measure the commonness of and N is the number of documents (d) in the corpus (D).. The denominator is simply the number of documents in which the term, t, appears in. 

IDF algorithm: idf(t, D) = log( N / count(d ∈ D : t ∈ d) )
!['IDF'](https://ecm.capitalone.com/WCM/tech/tf-idf-1.png)

Note: It can be possible for a term to not appear in the corpus at all, which can result in a divide-by-zero error. One way to handle this is to take the existing count and add 1. Thus making the denominator (1 + count). An example of how the  popular library scikit-learn handles this can be seen below.

Image with the sci-kit learn IDF algo, IDF(t) = log( (1+n) / (1 + df(t))) + 1 vs Standard notation idf algo IDF(t) = log(n/df(t))

!['IDF'](https://ecm.capitalone.com/WCM/tech/tf-idf-2.png)

The reason we need IDF is to help correct for words like “of”, “as”, “the”, etc. since they appear frequently in an English corpus. Thus by taking inverse document frequency, we can minimize the weighting of frequent terms while making infrequent terms have a higher impact.

Finally IDFs can also be pulled from either a background corpus, which corrects for sampling bias, or the dataset being used in the experiment at hand.

Putting it together: TF-IDF
To summarize the key intuition motivating TF-IDF is the importance of a term is inversely related to its frequency across documents.TF gives us information on how often a term appears in a document and IDF gives us information about the relative rarity of a term in the collection of documents. By multiplying these values together we can get our final TF-IDF value.


TF-IDF algo: tf-idf(t,d,D) = tf(t,d) x idf(t,d)
!['TF-IDF](https://ecm.capitalone.com/WCM/tech/tf-idf-3.png)

The higher the TF-IDF score the more important or relevant the term is; as a term gets less relevant, its TF-IDF score will approach 0.

In [112]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf_idf = TfidfVectorizer()
x = tf_idf.fit_transform(corpus).toarray()

In [113]:
y = df_combined.iloc[:,1].values

In [114]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.20 , random_state = 0)

In [115]:
from sklearn.naive_bayes import MultinomialNB
fake_news_model_2 = MultinomialNB().fit(x_train,y_train)

In [116]:
y_pred = fake_news_model_2.predict(x_test)

In [117]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test,y_pred)
matrix

array([[3304,  271],
       [ 277, 3894]], dtype=int64)

In [118]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.9292538084172476

Here, We got an accuracy of `92.9`%.This may change because using Random Sampling the highest accuracy i got was `94`%, Don't Worry it won't be much Different.

## Saving The Model
Saving Machine Learning Model using joblib.

In [120]:
filename2 = 'TF-IDF_model.sav'
joblib.dump(fake_news_model, filename2)

['TF-IDF_model.sav']

### To load the data:

loaded_model = joblib.load(filename2)
